src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,276 @@
+  private static final Meter          sendBucketErrorMeter = metricRegistry.meter(name(DirectoryReconciler.class, "sendBucketError"));
+
+  private static final String   WORKERS_KEY   = "directory_reconciliation_workers";
+  private static final String   COUNTER_KEY   = "directory_reconciliation_counter";
+  private static final long     WORKER_TTL_MS = 120_000L;
+  private static final long     BUCKET_ORDER  = 11L;
+  private static final long     BUCKET_COUNT  = 1L << BUCKET_ORDER;
+  private static final double   JITTER_MAX    = 0.20;
+  private static final String[] COUNTER_ARGS  = {
+          "OVERFLOW", "WRAP",
+          "INCRBY", "u32", "0", "1",
+          };
+
+  private final String              serverApiUrl;
+  private final ReplicatedJedisPool jedisPool;
+  private final AccountsManager     accountsManager;
+  private final Client              client;
+
+  private DirectoryReconciliationWorker directoryReconciliationWorker;
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
+                             ReplicatedJedisPool jedisPool,
+                             AccountsManager accountsManager)
+  {
+    this.serverApiUrl = cdsConfig.getServerApiUrl();
+    this.jedisPool = jedisPool;
+    this.accountsManager = accountsManager;
+
+    SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                               .securityProtocol("TLSv1.2");
+    try {
+      sslConfig.trustStore(initializeKeyStore(cdsConfig.getServerApiCaCertificate()));
+    } catch (CertificateException ex) {
+      logger.error("error reading serverApiCaCertificate from contactDiscovery config", ex);
+    }
+
+    this.client = ClientBuilder.newBuilder()
+                               .register(HttpAuthenticationFeature.basic("signal", cdsConfig.getServerApiToken().getBytes()))
+                               .sslContext(sslConfig.createSSLContext())
+                               .build();
+  }
+
+  private static KeyStore initializeKeyStore(String pemCaCertificate)
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public void start() {
+    this.directoryReconciliationWorker = new DirectoryReconciliationWorker(serverApiUrl, jedisPool, accountsManager, client);
```

Comment:
Let's just have the DirectoryReconciler implement Runnable?

Reply:
done in 2d295f73c5cd75724fb709f2cf5dfdaca4f6573e